pattern/pattern.c

```diff
@@ -434,65 +430,58 @@ int mutt_pattern_func(struct MailboxView *mv, int op, char *prompt)
 
 /**
  * mutt_search_command - Perform a search
- * @param mv   Mailbox view to search through
- * @param menu Current Menu
- * @param cur  Index number of current email
- * @param op   Operation to perform, e.g. OP_SEARCH_NEXT
- * @retval >=0 Index of matching email
- * @retval -1  No match, or error
+ * @param mv    Mailbox view to search through
+ * @param menu  Current Menu
+ * @param cur   Index number of current email
+ * @param state Current search state
+ * @param flags Search flags, e.g. SEARCH_PROMPT
+ * @retval >=0  Index of matching email
+ * @retval -1   No match, or error
  */
-int mutt_search_command(struct MailboxView *mv, struct Menu *menu, int cur, int op)
+int mutt_search_command(struct MailboxView *mv, struct Menu *menu, int cur,
+                        struct SearchState *state, SearchFlags flags)
 {
   struct Progress *progress = NULL;
-  struct Buffer *buf = NULL;
   int rc = -1;
   struct Mailbox *m = mv ? mv->mailbox : NULL;
   if (!m)
     return -1;
 
-  if ((*LastSearch == '\0') || ((op != OP_SEARCH_NEXT) && (op != OP_SEARCH_OPPOSITE)))
+  if (buf_is_empty(state->string) || (flags & SEARCH_PROMPT))
   {
-    buf = buf_pool_get();
-    buf_strcpy(buf, (LastSearch[0] != '\0') ? LastSearch : "");
-    if ((mw_get_field(((op == OP_SEARCH) || (op == OP_SEARCH_NEXT)) ? _("Search for: ") : _("Reverse search for: "),
-                      buf, MUTT_COMP_CLEAR | MUTT_COMP_PATTERN, false, NULL, NULL, NULL) != 0) ||
-        buf_is_empty(buf))
+    if ((mw_get_field((state->reverse) ? _("Reverse search for: ") : _("Search for: "),
+                      state->string, MUTT_COMP_CLEAR | MUTT_COMP_PATTERN, false,
+                      NULL, NULL, NULL) != 0) ||
+        buf_is_empty(state->string))
     {
       goto done;
     }
 
-    if ((op == OP_SEARCH) || (op == OP_SEARCH_NEXT))
-      OptSearchReverse = false;
-    else
-      OptSearchReverse = true;
-
     /* compare the *expanded* version of the search pattern in case
      * $simple_search has changed while we were searching */
     struct Buffer *tmp = buf_pool_get();
-    buf_copy(tmp, buf);
+    buf_copy(tmp, state->string);
     const char *const c_simple_search = cs_subset_string(NeoMutt->sub, "simple_search");
     mutt_check_simple(tmp, NONULL(c_simple_search));
 
-    if (!SearchPattern || !mutt_str_equal(buf_string(tmp), LastSearchExpn))
+    if (!state->pattern || !buf_str_equal(tmp, state->string_expn))
     {
       struct Buffer err;
       buf_init(&err);
       OptSearchInvalid = true;
-      mutt_str_copy(LastSearch, buf_string(buf), sizeof(LastSearch));
-      mutt_str_copy(LastSearchExpn, buf_string(tmp), sizeof(LastSearchExpn));
+      buf_copy(state->string_expn, tmp);
       mutt_message(_("Compiling search pattern..."));
-      mutt_pattern_free(&SearchPattern);
+      mutt_pattern_free(&state->pattern);
       err.dsize = 256;
       err.data = mutt_mem_malloc(err.dsize);
```

Comment:
Unrelated to my changes but do you know why we manually allocate the buffer data here? Shouldn't the `buf_*` functions simply allocate/grow the buffer if necessary?

Reply:
In the early days, `buf_new()` didn't exist, nor did the Buffer Pool.
In some code where the Buffer might not be needed (and to save an alloc) it was allocated on the stack.

With modern processors and lots of memory, we shouldn't be worrying about `malloc()`ing a few KB.
An interesting question is: Is the Buffer Pool really a saving?

Reply:
> An interesting question is: Is the Buffer Pool really a saving?

For the `string`, `string_expn` in the `SearchState` or in general?

Reply:
> With modern processors and lots of memory, we shouldn't be worrying about malloc()ing a few KB.

agreed, I've changed it

Reply:
> > Is the Buffer Pool really a saving?
> 
> For the `string`, `string_expn` in the `SearchState` or in general?

In general.
The _idea_ of a pool sounds good, but there aren't any _time-critical_ loops where we alloc/free Buffers.

> Shouldn't the `buf_*` functions simply allocate/grow the buffer if necessary?

Many of them don't: `if (!buf || !buf->data) FAIL;`
We just need a plan.

Reply:
> > > Is the Buffer Pool really a saving? 
> > 
> > For the `string`, `string_expn` in the `SearchState` or in general?
> 
> In general. The _idea_ of a pool sounds good, but there aren't any _time-critical_ loops where we alloc/free Buffers.

I can't really judge this. But I'm happy to rip it out if we decide to do it. ;-)